File: trunk/bin/split_file.pl
#!/usr/bin/perl -w
use strict;

if (@ARGV<4){
  print "<basefile> <partsize> <header=1|noheader=0> <is_haploid>\n";
  exit(1);
}

my $basefile=shift;
my $max=shift;
my $is_header=shift;
my $is_haploid=shift;
#my $sexfile=shift;

# read in the sexes
#my @sexarr=();
#open(IN,$sexfile)||die "Can't find $sexfile\n";
#<IN>;
#while(my $line=<IN>){
#  my ($seq,$sex)=split(/\t/,$line);
#  push(@sexarr,$sex);
#}
#close(IN);

my $i=0;
my $c=0;
my $header;
if ($is_header){
  $header=<STDIN>;
  chomp($header);
}
while(my $line=<STDIN>){
  chomp($line);
  if ($i % $max==0){
    if ($i){
      close(OUT3);
    }
    ++$c;
    my $outfile3=$basefile.$c;
    open(OUT3,">$outfile3");
    print STDERR "Writing to $outfile3\n";
    if ($is_header){
      print OUT3 "$header\n";
    }
  }
  #my $persons=scalar(@sexarr);
  my $persons=length($line);
  for(my $i=0;$i<$persons;++$i){
    my $c=substr($line,$i,1);
    #my $sex=$sexarr[$i];
    #if ($chr eq "X" && $sex eq "M" && $c eq "2"){
    if ($is_haploid && $c eq "2"){
      $c="3";
    }
    print OUT3 "$c";
  }
  print OUT3 "\n";
  ++$i;
}
close(OUT3);

File: branches/bam_support/src/utility/split_file.cpp
using namespace std;

#include<iostream>
#include<fstream>
#include<sstream>
#include<cstdlib>
#include<vector>

int main(int argc,char * argv[]){
  if (argc<4){
    cerr<<"Usage: <basefile> <partsize> <header=1|noheader=0>\n";
    exit(0);
  }
  int arg=0;
  const char * basefile = argv[++arg];
  int max = atoi(argv[++arg]);
  bool is_header = atoi(argv[++arg]);
  string line;
  int i=0,c=0;
  string header;
  if (is_header){
    getline(cin,header);
  }
  ofstream ofs;
  string outfile;
  while(getline(cin,line)){
    if ((i % max) == 0){
      if (i){
        ofs.close();
      }
      ++c;
      ostringstream oss;
      oss<<basefile<<c;
      outfile = oss.str();
      ofs.open(outfile.data());
      cerr<<"Writing to "<<outfile<<endl;
      if (is_header){
        ofs<<header<<endl;
      }
    }
    ofs<<line<<endl;
    ++i;
  }
  ofs.close();
}
